File: gatepack/frontend/expr.py
"""Boolean expression language for ``design.yaml`` (§10.2).

One language serves guards (transition ``when``), ``expressions``, ``output_logic``
and macro ``enable`` fields:

* identifiers reference inputs or named expressions,
* ``!`` (not), ``&`` (and), ``|`` (or), ``^`` (xor), parentheses,
* constants ``0`` / ``1``,
* ``state == NAME`` — a state-equality test used in ``output_logic`` and macro
  enables (Moore-style).

Precedence (tightest first): ``!``, ``&``, ``^``, ``|`` — matching
``gatepack.liberty.boolean``.  Emission fully parenthesises to remove any
precedence ambiguity.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping


class ExprError(ValueError):
    """An expression is malformed or invalid for its context."""


@dataclass(frozen=True)
class Var:
    name: str


@dataclass(frozen=True)
class Const:
    value: bool


@dataclass(frozen=True)
class Not:
    x: "Expr"


@dataclass(frozen=True)
class Bin:
    op: str
    left: "Expr"
    right: "Expr"


@dataclass(frozen=True)
class StateEq:
    state: str


Expr = Var | Const | Not | Bin | StateEq


_OPERATORS = {"&", "|", "^", "!", "(", ")", "=="}


def _tokenize(text: str) -> list[str]:
    tokens: list[str] = []
    i = 0
    n = len(text)
    while i < n:
        c = text[i]
        if c.isspace():
            i += 1
            continue
        if c in "&|^!()":
            tokens.append(c)
            i += 1
            continue
        if c == "=":
            if i + 1 < n and text[i + 1] == "=":
                tokens.append("==")
                i += 2
                continue
            raise ExprError(f"unexpected '=' in expression {text!r} (use '==')")
        if c.isdigit():
            j = i
            while j < n and text[j].isdigit():
                j += 1
            tokens.append(text[i:j])
            i = j
            continue
        if c.isalpha() or c == "_":
            j = i
            while j < n and (text[j].isalnum() or text[j] == "_"):
                j += 1
            tokens.append(text[i:j])
            i = j
            continue
        raise ExprError(f"unexpected character {c!r} in expression {text!r}")
    return tokens


def _parse_or(tokens: list[str], pos: int) -> tuple[Expr, int]:
    node, pos = _parse_xor(tokens, pos)
    while pos < len(tokens) and tokens[pos] == "|":
        rhs, pos = _parse_xor(tokens, pos + 1)
        node = Bin("|", node, rhs)
    return node, pos


def _parse_xor(tokens: list[str], pos: int) -> tuple[Expr, int]:
    node, pos = _parse_and(tokens, pos)
    while pos < len(tokens) and tokens[pos] == "^":
        rhs, pos = _parse_and(tokens, pos + 1)
        node = Bin("^", node, rhs)
    return node, pos


def _parse_and(tokens: list[str], pos: int) -> tuple[Expr, int]:
    node, pos = _parse_not(tokens, pos)
    while pos < len(tokens) and tokens[pos] == "&":
        rhs, pos = _parse_not(tokens, pos + 1)
        node = Bin("&", node, rhs)
    return node, pos


def _parse_not(tokens: list[str], pos: int) -> tuple[Expr, int]:
    if pos < len(tokens) and tokens[pos] == "!":
        node, pos = _parse_not(tokens, pos + 1)
        return Not(node), pos
    return _parse_atom(tokens, pos)


def _parse_atom(tokens: list[str], pos: int) -> tuple[Expr, int]:
    if pos >= len(tokens):
        raise ExprError("unexpected end of expression")
    token = tokens[pos]
    if token == "(":
        node, pos = _parse_or(tokens, pos + 1)
        if pos >= len(tokens) or tokens[pos] != ")":
            raise ExprError("missing ')' in expression")
        return node, pos + 1
    if token in _OPERATORS:
        raise ExprError(f"unexpected {token!r} in expression")
    if token.isdigit():
        if token not in ("0", "1"):
            raise ExprError(f"constant {token!r} must be 0 or 1")
        return Const(token == "1"), pos + 1
    if token == "state":
        if pos + 2 < len(tokens) and tokens[pos + 1] == "==" and tokens[pos + 2].isidentifier():
            return StateEq(tokens[pos + 2]), pos + 3
        raise ExprError("'state' may only appear as 'state == NAME'")
    return Var(token), pos + 1


def parse(text: str) -> Expr:
    """Parse an expression string into an AST."""
    tokens = _tokenize(text)
    if not tokens:
        raise ExprError("empty expression")
    node, pos = _parse_or(tokens, 0)
    if pos != len(tokens):
        raise ExprError(f"unexpected trailing tokens {tokens[pos:]!r} in {text!r}")
    return node


def free_vars(expr: Expr) -> set[str]:
    """Names of variables (inputs or named expressions) referenced by ``expr``."""
    if isinstance(expr, Var):
        return {expr.name}
    if isinstance(expr, Const):
        return set()
    if isinstance(expr, Not):
        return free_vars(expr.x)
    if isinstance(expr, Bin):
        return free_vars(expr.left) | free_vars(expr.right)
    if isinstance(expr, StateEq):
        return set()
    raise TypeError(expr)


def states_referenced(expr: Expr) -> set[str]:
    """State names compared against with ``state == NAME``."""
    if isinstance(expr, StateEq):
        return {expr.state}
    if isinstance(expr, (Var, Const)):
        return set()
    if isinstance(expr, Not):
        return states_referenced(expr.x)
    if isinstance(expr, Bin):
        return states_referenced(expr.left) | states_referenced(expr.right)
    raise TypeError(expr)


def expand(expr: Expr, definitions: Mapping[str, Expr]) -> Expr:
    """Substitute named-expression references with their definitions.

    Detects cycles (an expression defined in terms of itself, directly or
    transitively).
    """

    def go(e: Expr, seen: tuple[str, ...]) -> Expr:
        if isinstance(e, Var):
            if e.name in definitions:
                if e.name in seen:
                    chain = " -> ".join((*seen, e.name))
                    raise ExprError(f"cyclic expression definition: {chain}")
                return go(definitions[e.name], (*seen, e.name))
            return e
        if isinstance(e, Const):
            return e
        if isinstance(e, Not):
            return Not(go(e.x, seen))
        if isinstance(e, Bin):
            return Bin(e.op, go(e.left, seen), go(e.right, seen))
        if isinstance(e, StateEq):
            return e
        raise TypeError(e)

    return go(expr, ())


def evaluate(expr: Expr, env: Mapping[str, bool]) -> bool:
    """Evaluate a combinational expression under an input assignment.

    Raises ``ExprError`` if ``expr`` contains a ``state ==`` test (state is not
    an input).
    """
    if isinstance(expr, Var):
        if expr.name not in env:
            raise ExprError(f"unknown signal {expr.name!r} in evaluation")
        return env[expr.name]
    if isinstance(expr, Const):
        return expr.value
    if isinstance(expr, Not):
        return not evaluate(expr.x, env)
    if isinstance(expr, Bin):
        left = evaluate(expr.left, env)
        right = evaluate(expr.right, env)
        if expr.op == "&":
            return left and right
        if expr.op == "|":
            return left or right
        if expr.op == "^":
            return left != right
        raise ExprError(f"unknown operator {expr.op!r}")
    if isinstance(expr, StateEq):
        raise ExprError("state comparison has no combinational value")
    raise TypeError(expr)


def to_verilog(
    expr: Expr,
    var_map: Mapping[str, str],
    state_map: Mapping[str, str],
) -> str:
    """Emit a Verilog expression.

    ``var_map`` names inputs and named expressions; ``state_map`` maps a state
    name to the Verilog signal that is asserted when the FSM is in that state
    (the state bit for one-hot, ``(state == CODE)`` otherwise).
    """
    if isinstance(expr, Var):
        if expr.name not in var_map:
            raise ExprError(f"unknown signal {expr.name!r} in Verilog emission")
        return var_map[expr.name]
    if isinstance(expr, Const):
        return "1'b1" if expr.value else "1'b0"
    if isinstance(expr, Not):
        return f"(~{to_verilog(expr.x, var_map, state_map)})"
    if isinstance(expr, Bin):
        op = {"&": "&", "|": "|", "^": "^"}[expr.op]
        return (
            f"({to_verilog(expr.left, var_map, state_map)} {op} "
            f"{to_verilog(expr.right, var_map, state_map)})"
        )
    if isinstance(expr, StateEq):
        if expr.state not in state_map:
            raise ExprError(f"unknown state {expr.state!r} in Verilog emission")
        return state_map[expr.state]
    raise TypeError(expr)

File: app/renderer/design/expr.ts
/**
 * Client-side boolean expression language for `design.yaml` (§10.2).
 *
 * This mirrors `gatepack/frontend/expr.py` exactly — the *same* language the
 * core parses — so the live truth-table column (the sanctioned client-side
 * exception) evaluates identically to the core:
 *
 *   - identifiers reference inputs or named expressions,
 *   - `!` (not), `&` (and), `|` (or), `^` (xor), parentheses,
 *   - constants `0` / `1`,
 *   - `state == NAME` (Moore-style).
 *
 * Precedence (tightest first): `!`, `&`, `^`, `|`.
 *
 * Evaluation is tri-valued: `'0'`, `'1'`, or `'x'` for unknown (a don't-care
 * input, or a `state ==` test with no state context). This is what lets the
 * truth table render don't-cares honestly instead of guessing.
 */

export class ExprError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ExprError';
  }
}

export type Bit = '0' | '1' | 'x';

export type Expr =
  | { kind: 'var'; name: string }
  | { kind: 'const'; value: boolean }
  | { kind: 'not'; x: Expr }
  | { kind: 'bin'; op: '&' | '|' | '^'; left: Expr; right: Expr }
  | { kind: 'stateEq'; state: string };

export type Env = Record<string, Bit>;

/* ------------------------------------------------------------------ */
/* Tokeniser                                                           */
/* ------------------------------------------------------------------ */

function tokenize(text: string): string[] {
  const tokens: string[] = [];
  let i = 0;
  const n = text.length;
  while (i < n) {
    const c = text[i];
    if (/\s/.test(c)) {
      i += 1;
      continue;
    }
    if ('&|^!()'.includes(c)) {
      tokens.push(c);
      i += 1;
      continue;
    }
    if (c === '=') {
      if (i + 1 < n && text[i + 1] === '=') {
        tokens.push('==');
        i += 2;
        continue;
      }
      throw new ExprError(`unexpected '=' in expression ${JSON.stringify(text)} (use '==')`);
    }
    if (c >= '0' && c <= '9') {
      let j = i;
      while (j < n && text[j] >= '0' && text[j] <= '9') j += 1;
      tokens.push(text.slice(i, j));
      i = j;
      continue;
    }
    if (/[A-Za-z_]/.test(c)) {
      let j = i;
      while (j < n && /[A-Za-z0-9_]/.test(text[j])) j += 1;
      tokens.push(text.slice(i, j));
      i = j;
      continue;
    }
    throw new ExprError(`unexpected character ${JSON.stringify(c)} in expression ${JSON.stringify(text)}`);
  }
  return tokens;
}

/* ------------------------------------------------------------------ */
/* Parser (recursive descent, precedence ! > & > ^ > |)                */
/* ------------------------------------------------------------------ */

function parseOr(tokens: string[], pos: number): [Expr, number] {
  let [node, p] = parseXor(tokens, pos);
  while (p < tokens.length && tokens[p] === '|') {
    const [rhs, q] = parseXor(tokens, p + 1);
    node = { kind: 'bin', op: '|', left: node, right: rhs };
    p = q;
  }
  return [node, p];
}

function parseXor(tokens: string[], pos: number): [Expr, number] {
  let [node, p] = parseAnd(tokens, pos);
  while (p < tokens.length && tokens[p] === '^') {
    const [rhs, q] = parseAnd(tokens, p + 1);
    node = { kind: 'bin', op: '^', left: node, right: rhs };
    p = q;
  }
  return [node, p];
}

function parseAnd(tokens: string[], pos: number): [Expr, number] {
  let [node, p] = parseNot(tokens, pos);
  while (p < tokens.length && tokens[p] === '&') {
    const [rhs, q] = parseNot(tokens, p + 1);
    node = { kind: 'bin', op: '&', left: node, right: rhs };
    p = q;
  }
  return [node, p];
}

function parseNot(tokens: string[], pos: number): [Expr, number] {
  if (pos < tokens.length && tokens[pos] === '!') {
    const [node, p] = parseNot(tokens, pos + 1);
    return [{ kind: 'not', x: node }, p];
  }
  return parseAtom(tokens, pos);
}

function parseAtom(tokens: string[], pos: number): [Expr, number] {
  if (pos >= tokens.length) {
    throw new ExprError('unexpected end of expression');
  }
  const token = tokens[pos];
  if (token === '(') {
    const [node, p] = parseOr(tokens, pos + 1);
    if (p >= tokens.length || tokens[p] !== ')') {
      throw new ExprError("missing ')' in expression");
    }
    return [node, p + 1];
  }
  if ('&|^!()=='.includes(token)) {
    throw new ExprError(`unexpected ${JSON.stringify(token)} in expression`);
  }
  if (/^[0-9]+$/.test(token)) {
    if (token !== '0' && token !== '1') {
      throw new ExprError(`constant ${JSON.stringify(token)} must be 0 or 1`);
    }
    return [{ kind: 'const', value: token === '1' }, pos + 1];
  }
  if (token === 'state') {
    if (pos + 2 < tokens.length && tokens[pos + 1] === '==' && /^[A-Za-z_][A-Za-z0-9_]*$/.test(tokens[pos + 2])) {
      return [{ kind: 'stateEq', state: tokens[pos + 2] }, pos + 3];
    }
    throw new ExprError("'state' may only appear as 'state == NAME'");
  }
  return [{ kind: 'var', name: token }, pos + 1];
}

export function parse(text: string): Expr {
  const tokens = tokenize(text);
  if (tokens.length === 0) {
    throw new ExprError('empty expression');
  }
  const [node, pos] = parseOr(tokens, 0);
  if (pos !== tokens.length) {
    throw new ExprError(`unexpected trailing tokens ${JSON.stringify(tokens.slice(pos))} in ${JSON.stringify(text)}`);
  }
  return node;
}

/* ------------------------------------------------------------------ */
/* Analysis                                                            */
/* ------------------------------------------------------------------ */

export function freeVars(expr: Expr): Set<string> {
  switch (expr.kind) {
    case 'var':
      return new Set([expr.name]);
    case 'const':
      return new Set();
    case 'not':
      return freeVars(expr.x);
    case 'bin':
      return new Set([...freeVars(expr.left), ...freeVars(expr.right)]);
    case 'stateEq':
      return new Set();
  }
}

export function statesReferenced(expr: Expr): Set<string> {
  switch (expr.kind) {
    case 'stateEq':
      return new Set([expr.state]);
    case 'var':
    case 'const':
      return new Set();
    case 'not':
      return statesReferenced(expr.x);
    case 'bin':
      return new Set([...statesReferenced(expr.left), ...statesReferenced(expr.right)]);
  }
}

/**
 * Substitute named-expression references with their definitions, detecting
 * cycles (an expression defined in terms of itself, directly or transitively).
 */
export function expand(expr: Expr, definitions: Record<string, Expr>): Expr {
  function go(e: Expr, seen: string[]): Expr {
    switch (e.kind) {
      case 'var': {
        if (e.name in definitions) {
          if (seen.includes(e.name)) {
            throw new ExprError(`cyclic expression definition: ${[...seen, e.name].join(' -> ')}`);
          }
          return go(definitions[e.name], [...seen, e.name]);
        }
        return e;
      }
      case 'const':
      case 'stateEq':
        return e;
      case 'not':
        return { kind: 'not', x: go(e.x, seen) };
      case 'bin':
        return { kind: 'bin', op: e.op, left: go(e.left, seen), right: go(e.right, seen) };
    }
  }
  return go(expr, []);
}

/* ------------------------------------------------------------------ */
/* Evaluation (tri-valued)                                             */
/* ------------------------------------------------------------------ */

function not(b: Bit): Bit {
  if (b === 'x') return 'x';
  return b === '1' ? '0' : '1';
}

function and(a: Bit, b: Bit): Bit {
  if (a === '0' || b === '0') return '0';
  if (a === 'x' || b === 'x') return 'x';
  return '1';
}

function or(a: Bit, b: Bit): Bit {
  if (a === '1' || b === '1') return '1';
  if (a === 'x' || b === 'x') return 'x';
  return '0';
}

function xor(a: Bit, b: Bit): Bit {
  if (a === 'x' || b === 'x') return 'x';
  return a === b ? '0' : '1';
}

/**
 * Evaluate an expression under an input assignment. `state` is the optional
 * state context for `state == NAME` tests; when omitted such a test is `'x'`.
 */
export function evaluate(expr: Expr, env: Env, state?: string): Bit {
  switch (expr.kind) {
    case 'var':
      return expr.name in env ? env[expr.name] : 'x';
    case 'const':
      return expr.value ? '1' : '0';
    case 'not':
      return not(evaluate(expr.x, env, state));
    case 'bin': {
      const left = evaluate(expr.left, env, state);
      const right = evaluate(expr.right, env, state);
      if (expr.op === '&') return and(left, right);
      if (expr.op === '|') return or(left, right);
      return xor(left, right);
    }
    case 'stateEq':
      return state === undefined ? 'x' : state === expr.state ? '1' : '0';
  }
}
